datafusion/core/src/physical_plan/hash_join.rs

```diff
@@ -791,6 +826,109 @@ fn build_join_indexes(
     }
 }
 
+fn apply_join_filter(
+    left: &RecordBatch,
+    right: &RecordBatch,
+    join_type: JoinType,
+    left_indices: UInt64Array,
+    right_indices: UInt32Array,
+    filter: &JoinFilter,
+) -> Result<(UInt64Array, UInt32Array)> {
+    if left_indices.is_empty() {
+        return Ok((left_indices, right_indices));
+    };
+
+    let (intermediate_batch, _) = build_batch_from_indices(
+        filter.schema(),
+        left,
+        right,
+        PrimitiveArray::from(left_indices.data().clone()),
+        PrimitiveArray::from(right_indices.data().clone()),
+        filter.column_indices(),
+    )?;
+
+    match join_type {
+        JoinType::Inner | JoinType::Left => {
+            // For both INNER and LEFT joins, input arrays contains only indices for matched data.
```

Comment:
Something bothers me about this code treating `Left` and `Right` joins differently -- I would expect they look like mirrors of each other and `Full` to be treated differently

Reply:
That was my initial plan, but it turned out there are [no nulls appended](https://github.com/apache/arrow-datafusion/blob/8e71c887621254b3d18523ad182045a9e987ee12/datafusion/core/src/physical_plan/hash_join.rs#L728) in case of left join while building matched indices. All null values are handled [later](https://github.com/apache/arrow-datafusion/blob/8e71c887621254b3d18523ad182045a9e987ee12/datafusion/core/src/physical_plan/hash_join.rs#L1024), using visited left side indices